watertap/unit_models/crystallizer.py

```diff
@@ -0,0 +1,696 @@
+
+        # (ii) Fix liquids: solid and vapour flows must be zero
+        @self.Constraint(self.properties_out[0].phase_component_set,
+            doc="Empty components in liquid stream")
+        def eq_liquid_composition(b, p, j):
+            if p != 'Liq':
+                return (b.properties_out[0].flow_mass_phase_comp[p, j] == 1e-8 * pyunits.kg/pyunits.s)
+            else:
+                return Constraint.Skip
+
+        # (iii) Fix vapor: solid and vapour flows must be zero. Vapour is also pure solvent.
+        @self.Constraint(self.properties_vapor[0].phase_component_set,
+            doc="Empty components in vapour stream")
+        def eq_vapor_composition(b, p, j):
+            if p != 'Vap':
+                return (b.properties_vapor[0].flow_mass_phase_comp[p, j] == 1e-8 * pyunits.kg/pyunits.s)
+            else:
+                return Constraint.Skip
+
+
+        # 5. Add an energy balance for the system
+        ## (iii) Enthalpy balance: based on Lewis et al. Enthalpy is exothermic, and the hC in property package is -ve
+        @self.Constraint(doc="Enthalpy balance over crystallization system")
+        def eq_enthalpy_balance(b):
+            return (b.properties_in[0].enth_flow 
+                - b.properties_out[0].enth_flow - b.properties_vapor[0].enth_flow - b.properties_solids[0].enth_flow
+                + self.work_mechanical[0]
+                - (b.properties_solids[0].flow_vol_phase['Sol'] * b.properties_solids[0].dens_mass_solute['Sol'] * b.properties_solids[0].dh_crystallization['NaCl'])
+                == 0
+                )
+
+
+        # 6. Pressure and temperature balances - what is the pressure of the outlet solid and vapour?
+        # TO-DO: Figure out actual liquid and solid pressures.
+        @self.Constraint()    
+        def eq_p_con1(b):
+            return b.properties_in[0].pressure - b.properties_out[0].pressure == 0
+
+        @self.Constraint()    
+        def eq_p_con2(b):
+            return b.properties_in[0].pressure - b.properties_solids[0].pressure == 0
+
+        @self.Constraint()    
+        def eq_p_con3(b):
+            return b.properties_vapor[0].pressure - self.pressure_operating == 0
+
+        @self.Constraint()    
+        def eq_T_con1(b):
+            return self.temperature_operating - b.properties_solids[0].temperature == 0
+
+        @self.Constraint()    
+        def eq_T_con2(b):
+            return self.temperature_operating - b.properties_vapor[0].temperature == 0
+
+        @self.Constraint()
+        def eq_T_con3(b):
+            return self.temperature_operating - b.properties_out[0].temperature == 0
+
+
+        # 7. Heat exchanger minimum circulation flow rate calculations - see Lewis et al. or Tavare et al.
+        @self.Constraint(doc="Constraint on mimimum circulation rate through crystallizer heat exchanger")
+        def eq_minimum_hex_circulation_rate_constraint(b):
+            rho_cp_dt = self.hex_max_temp_increase * \
```

Comment:
I'd write out `rho` as `dens`, although I'm not 100% understanding this variable. It seems to be the average of the product of density and cp. I think `dens_cp_avg` could work. It's also minor because this is just a temporary expression, but might as well be consistent.

Reply:
Yes, it is a weighted average of the sum of the product of density and cp for the solid/liquid mixture. Name changed.